tests/test_parquet.py

```diff
@@ -0,0 +1,116 @@
+# for details of code ownership.
+#
+# This program is free software: you can redistribute it and/or modify
+# it under the terms of the GNU General Public License as published by
+# the Free Software Foundation, either version 3 of the License, or
+# (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+"""Tests for ParquetFormatter.
+
+Tests in this module are disabled unless pandas and pyarrow are importable.
+"""
+
+import os
+import unittest
+import tempfile
+import shutil
+
+try:
+    import numpy as np
+    import pandas as pd
+    import pyarrow
+except ImportError:
+    pyarrow = None
+
+from lsst.daf.butler import Butler, DatasetType
+
+
+TESTDIR = os.path.abspath(os.path.dirname(__file__))
+
+
+@unittest.skipUnless(pyarrow is not None, "Cannot tests ParquetFormatter without pyarrow.")
+class ParquetFormatterTestCase(unittest.TestCase):
+    """Tests for ParquetFormatter, using PosixDatastore.
+    """
+
+    def setUp(self):
+        """Create a new butler root for each test."""
+        self.root = tempfile.mkdtemp(dir=TESTDIR)
+        Butler.makeRepo(self.root)
+        self.butler = Butler(self.root, run="test_run")
+        # No dimensions in dataset type so we don't have to worry about
+        # inserting dimension data or defining data IDs.
+        self.datasetType = DatasetType("data", dimensions=(), storageClass="DataFrame",
+                                       universe=self.butler.registry.dimensions)
+        self.butler.registry.registerDatasetType(self.datasetType)
+
+    def tearDown(self):
+        if os.path.exists(self.root):
+            shutil.rmtree(self.root, ignore_errors=True)
+
+    def testSingleIndexDataFrame(self):
+        columns1 = pd.Index(["a", "b", "c"])
+        df1 = pd.DataFrame(np.random.randn(5, 3), index=np.arange(5, dtype=int), columns=columns1)
+        self.butler.put(df1, self.datasetType)
+        # Read the whole DataFrame.
+        df2 = self.butler.get(self.datasetType)
+        self.assertTrue(df1.equals(df2))
+        # Read just the column descriptions.
+        columns2 = self.butler.get(f"{self.datasetType.name}.columns")
+        self.assertTrue(df1.columns.equals(columns2))
+        # Read just some columns a few different ways.
+        df3 = self.butler.get(self.datasetType, parameters={"columns": ["a", "c"]})
+        self.assertTrue(df1.loc[:, ["a", "c"]].equals(df3))
+        df4 = self.butler.get(self.datasetType, parameters={"columns": "a"})
+        self.assertTrue(df1.loc[:, ["a"]].equals(df4))
+        # Passing an unrecognized column should be a ValueError.
```

Comment:
I think that exception is chained to the real one, i.e. it'll appear above the
```
The above exception was the direct cause of the following exception:
```
that precedes this one in the traceback.  I'm not a huge fan of that either, as it still obscures the more useful error message.  @timj, what do you think about having the catch-and-reraise at https://github.com/lsst/daf_butler/blob/80af4660f914ca6efcbfabd79678d63f3061b979/python/lsst/daf/butler/datastores/posixDatastore.py#L145 pass through exceptions that are already `ValueError` (or some other type) unchanged?

I've made a comment on DM-21976 and linked to this ticket.  Just to make sure we're on the same page, it's okay to leave the (different) behavior here as-is for now?

Reply:
My annoyance that `raise X from e` gives us unreadable stack traces knows no bounds. It's seemingly worse when run from pytest since when pytest reports the stack trace it stops at the boundary.

I'm not entirely sure what the best answer is. On the one hand we could let the formatter fail directly without catching it but then we can't document that you get a ValueError (you might get an ImportError for example) and it's nice that the trace does tell you what dataset you were trying to get and that it was the formatter that failed.

Are we doing it wrong? Is there some other syntax for raise that would work better?

Reply:
I think what we're doing is as right as it can be given language constraints.  Maybe in the kind of high-level contexts where we currently squash exception tracebacks (on the assumption that they're not what users want to see) we'd want to print the original message last and most prominently.  But I personally tend to find that behavior annoying anyway, because we haven't got good enough error messages for many failure modes, so usually it just means we tell the user to try again and pass `--doRaise` or something so we can actually help them.

Anyhow, doing:
```
try:
    ...
except ValueError:
    raise
except Exception as err:
    raise ValueError(...) from err
```
*would* still let the `put` interface guarantee a certain exception, but it would mean we don't pass along potentially useful information like the data ID and dataset type (which the formatter doesn't know).  So at this point I'm back to thinking what we have now is still the best option, even though none of them are good.